File: 프로그래머스/4/12929. 올바른 괄호의 갯수/올바른 괄호의 갯수.java
public class Solution {

    public static int answer = 0;

    public static int solution(int n) {
        func(0, 0, n);
        return answer;
    }

    public static void func(int left, int right, int n)
    {
        if (left == n && right == n)
        {
            ++answer;
            return;
        }
        if (left < n)
        {
            func(left + 1, right, n);
        }

        if (right < left)
        {
            func(left, right + 1, n);
        }
    }
}

File: 프로그래머스/4/12929. 올바른 괄호의 갯수/올바른 괄호의 갯수.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;

void dfs(int left, int right)
{
    if (left == 0 && right == 0)
    {
        ++answer;
    }

    if (left > 0)
        dfs(left - 1, right);
    
    if (left < right)
    {
        dfs(left, right - 1);
    }
}

int solution(int n) {
    dfs(n, n);
    return answer;
}
